### PR TITLE
Add QueueInput/OutputStream as simpler alternatives to PipedInput/OutputStream

### PR DESCRIPTION
**Pull request by maxxedev**
Add QueueInput/OutputStream as simpler alternatives to PipedInput/OutputStream

PipedInput/OutputStream in JDK can have surprisingly complex behavior with respect
to how threads need to be arranged. QueueInput/OutputStream are much simpler
alternatives that is easier to use correctly.

### CODE REVIEW

**Comment by garydgregory** [comment]
IMO, you need better tests with edge cases like input size of 0, 1, 2, 4, 8, and the "usual" buffer sizes from other streams like 4096 or the more recent 8192 (private java.io.BufferedInputStream.DEFAULT_BUFFER_SIZE).
You can parameterize the tests with normal JUnit features. Someone (like me ;-) will invariably look at the current implementation and not like the one byte at a time implementations since normal streams usually work on buffers, but that's a different story, let's start with better tests.

File: src/main/java/org/apache/commons/io/output/QueueOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.commons.io.input.QueueInputStream;
+
+import java.io.InterruptedIOException;
+import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import static java.util.Objects.requireNonNull;
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/output/QueueOutputStream.java:28]: We don't use static imports for this kind of call in the rest of the main code base. In unit tests, we use statics for JUnit methods.

File: src/main/java/org/apache/commons/io/output/QueueOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     *
+     * @throws InterruptedIOException if the thread is interrupted while writing to the queue.
+     */
+    @Override
+    public void write(final int b) throws InterruptedIOException {
+        try {
+            queue.put(0xFF & b);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new InterruptedIOException();
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/output/QueueOutputStream.java:92]: Is there a reason, the catch block creates a new exception instead of rethrowing the one it caught?

Comment (maxxedev) [reply src/main/java/org/apache/commons/io/output/QueueOutputStream.java:92]: InterruptedException does not extend IOException but InterruptedIOException does.

Comment (garydgregory) [reply src/main/java/org/apache/commons/io/output/QueueOutputStream.java:92]: Then we shouldn't we still preserve the original exception by calling `initCause()` on the new exception?

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     * @return QueueOutputStream connected to this stream
+     */
+    public QueueOutputStream newQueueOutputStream() {
+        return new QueueOutputStream(queue);
+    }
+
+    /**
+     * Read a single byte.
+     *
+     * @return either the byte read or <code>-1</code> if the end of the stream has been reached
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/input/QueueInputStream.java:82]: You can abbreviate ` <code>-1</code>` as `{@code -1}`.
You could also avoid using a magic number and do a static import of `IOUtils.EOF` in the code.

**Review by garydgregory — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ * 
+ * @see QueueOutputStream
+ * @since 2.9.0
+ */
+public class QueueInputStream extends InputStream {
+
+    private final BlockingQueue<Integer> queue;
+
+    /**
+     * Create a {@link QueueInputStream} with no limit to internal buffer size
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/input/QueueInputStream.java:55]: `Create` -> `Constructs` since a constructor constructs ;-)

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     * Create a {@link QueueInputStream} with given buffer
+     * 
+     * @param queue backing queue for the stream
+     */
+    public QueueInputStream(final BlockingQueue<Integer> queue) {
+        this.queue = Objects.requireNonNull(queue, "queue is required");
+    }
+
+    /**
+     * Create an QueueOutputStream connected to this. Writes to the output stream will be visible to this input stream.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/input/QueueInputStream.java:71]: `an QueueOutputStream` -> `an instance`.

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+     * 
+     * @return QueueOutputStream connected to this stream
+     */
+    public QueueOutputStream newQueueOutputStream() {
+        return new QueueOutputStream(queue);
+    }
+
+    /**
+     * Read a single byte.
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/input/QueueInputStream.java:80]: `Read` -> `Reads`.

File: src/main/java/org/apache/commons/io/output/QueueOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ * 
+ * @see QueueInputStream
+ * @since 2.9.0
+ */
+public class QueueOutputStream extends OutputStream {
+
+    private final BlockingQueue<Integer> queue;
+
+    /**
+     * Create a {@link QueueOutputStream} with no limit to internal buffer size
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/output/QueueOutputStream.java:57]: -> `Creates a new instance...` 

I am not a fan of creating a hyperlink to the class we are already in, it feels misleading, like we are sending someone elsewhere, when we are not.

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java (hunk 3 above)
Comment (maxxedev) [reply src/main/java/org/apache/commons/io/input/QueueInputStream.java:82]: -1 as magic number comes from [InputStream::read](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read--). I think it'd better to keep it that way to be consistent with super method documentation.

File: src/main/java/org/apache/commons/io/input/QueueInputStream.java (hunk 5 above)
Comment (maxxedev) [reply src/main/java/org/apache/commons/io/input/QueueInputStream.java:71]: This Queue**Input**Stream method creates a Queue**Output**Stream instance. Just saying "an instance" might be confusing?

Comment (garydgregory) [reply src/main/java/org/apache/commons/io/input/QueueInputStream.java]: Ah, yes of course!

**Comment by garydgregory** [comment]
I will review all of this tomorrow (Thursday).
